src/core/autodiff.cc

```diff
@@ -0,0 +1,387 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tc/core/autodiff.h"
+#include "tc/core/tc2halide.h"
+#include "tc/lang/parser.h"
+#include "tc/lang/sema.h"
+#include "tc/lang/tc_format.h"
+#include "tc/lang/tree_views.h"
+
+#include <unordered_set>
+#include <vector>
+
+namespace tc {
+
+using namespace lang;
+
+static const lang::SourceRange dummyRange{std::make_shared<std::string>(""),
+                                          0,
+                                          0};
+
+int32_t getTcType(Halide::Type t) {
+  if (t.is_int()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_INT64;
+      case 32:
+        return TK_INT32;
+      case 16:
+        return TK_INT16;
+      case 8:
+        return TK_INT8;
+    }
+  } else if (t.is_uint()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_UINT64;
+      case 32:
+        return TK_UINT32;
+      case 16:
+        return TK_UINT16;
+      case 8:
+        return TK_UINT8;
+    }
+  } else if (t.is_float()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_DOUBLE;
+      case 32:
+        return TK_FLOAT;
+    }
+  }
+  throw std::runtime_error("Unknown Halide type");
+}
+
+void findAccessedTensors(
```

Comment:
We have been implementing this type of functions as:
```
std::unordered_set<std::string> collectAccessedTensors(const TreeRef& tree) {...}
```
This way it can compose. 
Generally we try not to pass out parameters as mutable in but just return the result.